lib/connection.js

```diff
@@ -285,17 +287,31 @@ function Connection(options = {}) {
   }
 
   function data(x) {
-    buffer = buffer.length === 0
-      ? x
-      : Buffer.concat([buffer, x], buffer.length + x.length)
+    if (bs) {
+      bs.push(x)
+      rest -= x.length
+      if (rest >= 0)
+        return
+    }
+
+    buffer = bs
+      ? Buffer.concat(bs)
+      : buffer.length === 0
+        ? x
+        : Buffer.concat([buffer, x], buffer.length + x.length)
```

Comment:
Defining the length of the new Buffer (via `buffer.length + x.length`) is not necessary unless you want the new buffer to be larger/smaller than `buffer.length + x.length`

Reply:
I recall early on I did some perf testing on this, and it's faster because node will internally iterate the array to get the length.